apps/music/js/queue.js

```diff
@@ -0,0 +1,335 @@
+/* global asyncStorage, Database */
+/* exported PlaybackQueue */
+'use strict';
+
+var PlaybackQueue = (function() {
+
+  // The key for storing the repeat and shuffle options.
+  var SETTINGS_OPTION_KEY = 'settings_option_key';
+
+  var Repeat = {
+    OFF: 0,
+    LIST: 1,
+    SONG: 2,
+
+    /**
+     * Return the "next" repeat setting in its canonical order; OFF -> LIST ->
+     * SONG -> OFF.
+     *
+     * @param {Repeat} val The current value.
+     * @return {Repeat} The next value.
+     */
+    next: function(val) {
+      return (val + 1) % 3;
+    }
+  };
+
+  /**
+   * Return a % b, ensuring that the result is always positive.
+   *
+   * @param {Number} a The dividend.
+   * @param {Number} b The divisor.
+   * @return {Number} The positive remainder.
+   */
+  function posmod(a, b) {
+    var r = a % b;
+    return r < 0 ? r + b : r;
+  }
+
+  /**
+   * Fill an array with values from 0..n.
+   *
+   * @param {Number} length The length of the array.
+   * @return {Array} The array.
+   */
+  function fillIndices(length) {
+    var indices = new Array(length);
+    for (var i = 0; i < indices.length; i++) {
+      indices[i] = i;
+    }
+    return indices;
+  }
+
+  /**
+   * Shuffle the elements of an array in-place using Yates shuffle. See:
+   * <http://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle>.
+   *
+   * @param {Array} list The array to shuffle.
+   */
+  function shuffle(list) {
+    for (var i = list.length - 1; i >= 1; i--) {
+      var j = Math.floor(Math.random() * (i + 1));
+      if (j < i) {
+        var tmp = list[j];
+        list[j] = list[i];
+        list[i] = tmp;
+      }
+    }
+  }
+
+  /**
+   * The base for playback queue types.
+   */
+  function BaseQueue() {
+    this._index = null;
+  }
+
+  BaseQueue.prototype = {
```

Comment:
I actually saw somewhere recently that defining the `prototype` this way blows away the `constructor` property. You might want to just add to this:
```
  constructor: BaseQueue,
```